### PR TITLE
RELENG-4902 refactor upload_checks() (fixes, optimization, modularity)

### PR DESCRIPTION
**Pull request by bertranddemiddelaer**
refactor upload_checks() (fixes, optimization, modularity)

### CODE REVIEW

**Comment by bertranddemiddelaer** [comment]
Note that now, copy operations are restricted in the same way that was meant initially for upload. This is now consistent.

File: lua/github_access.lua [hunk 1]
```diff
@@ -7,25 +7,23 @@ local github_api_enabled            = os.getenv('GITHUB_API_ENABLED')
...
-    table.insert(github_restriction_upload, allowed_user)
-    ngx.log(ngx.STDERR, allowed_user)
+    table.insert(github_restriction_users, allowed_user)
 end
 
 function wrong_credentials()
    ngx.header.content_type = 'text/plain'
    ngx.header['WWW-Authenticate'] = 'Basic realm="Access to the Scality Artifacts", charset="UTF-8"'
    ngx.status = ngx.HTTP_UNAUTHORIZED
-   ngx.log(ngx.STDERR, '401 Access Denied')
```
Comment (bertranddemiddelaer) [inline lua/github_access.lua:21]: This log adds nothing as the status will already be logged by nginx

File: lua/github_access.lua [hunk 2]
```diff
@@ -7,25 +7,23 @@ local github_api_enabled            = os.getenv('GITHUB_API_ENABLED')
...
    return ngx.exit(ngx.HTTP_UNAUTHORIZED)
 end
 
 function not_allowed()
    ngx.header.content_type = 'text/plain'
    ngx.status = ngx.HTTP_FORBIDDEN
-   ngx.log(ngx.STDERR, '403 Forbidden response')
```
Comment (bertranddemiddelaer) [inline lua/github_access.lua:28]: Same here

File: lua/github_access.lua [hunk 3]
```diff
@@ -7,25 +7,23 @@ local github_api_enabled            = os.getenv('GITHUB_API_ENABLED')
 local github_api_company            = os.getenv('GITHUB_API_COMPANY')
 local env_github_restriction_upload = os.getenv('GITHUB_USER_ALLOWED_UPLOAD')
 local github_auth_cache_dir         = "/data/nginx/artifacts_github_auth_cache"
-local github_restriction_upload = {}
```
Comment (bertranddemiddelaer) [inline lua/github_access.lua:10]: use a more generic approach to restrict on upload AND copy (write operations are not protected in a consistent way).

File: lua/github_access.lua [hunk 4]
```diff
@@ -61,32 +60,33 @@ function update_cache(auth_md5, status)
...
+    local divider = auth:find(':')
     local username = auth:sub(0, divider-1)
     local auth_md5 = ngx.md5(auth)
 
-    cached_status = read_cache(auth_md5)
+    local cached_status = read_cache(auth_md5)
 
     if cached_status == nil then
         local res =  ngx.location.capture("/force_github_request/orgs/" .. github_api_company .. "/members/" .. username)
-        ngx.log(ngx.STDERR, "status member \t\t" .. res.status)
```
Comment (bertranddemiddelaer) [inline lua/github_access.lua:80]: no need to log a normal event, and ngx.STDERR would not be the place for that anyway

File: tests/.env [hunk 5]
```diff
@@ -7,4 +7,4 @@ ARTIFACTS_URL=http://artifacts
 GITHUB_API_URL=http://fake-github-url:5050
 GITHUB_API_COMPANY=scality
 GITHUB_API_ENABLED=true
-GITHUB_USER_ALLOWED_UPLOAD="username-pass"
```
Comment (bertranddemiddelaer) [inline tests/.env:10]: This one was really tricky ! The " chars were part of the value. So this made the tests script fail.

File: tests/github-fake-config/nginx.conf [hunk 6]
```diff
@@ -11,11 +11,11 @@ http {
     location /orgs/scality/members/username-pass {
         return 204;
     }
+    location /orgs/scality/members/username-pass-no-restricted-paths {
+        return 204;
+    }
     location /orgs/scality/members/username-fail {
         return 401;
     }
-    location /user {
```
Comment (bertranddemiddelaer) [inline tests/github-fake-config/nginx.conf:17]: useless, we do not use this github api call anymore

File: lua/github_access.lua [hunk 7]
```diff
@@ -61,32 +60,33 @@ function update_cache(auth_md5, status)
...
         update_cache(auth_md5, res.status)
 
         if res.status ~= 204 then
+            ngx.log(ngx.STDERR, 'User ' .. username .. ' not allowed (github auth cache MISS)')
             return  false
         end
         return true
     else
         if cached_status ~= '204' then
+            ngx.log(ngx.STDERR, 'User ' .. username .. ' not allowed (github auth cache HIT)')
```
Comment (bertranddemiddelaer) [inline lua/github_access.lua:89]: I though this one and the log above would be super useful if we want to know who refuses the access (cache or github)

File: lua/github_access.lua [hunk 8]
```diff
@@ -112,33 +112,44 @@ function verify_header()
...
+        if string.sub(ngx.var.request_uri, 1,string.len(v)) == v then
+            location = v
+            break
+        end
+    end
+
+    if location == nil then
         return true
     end
-    local res =  ngx.location.capture("/force_github_request/user")
```
Comment (bertranddemiddelaer) [inline lua/github_access.lua:120]: useless as we can capture the username with basic auth header.

File: tests/end2end/test_simple.py [hunk 9]
```diff
@@ -429,6 +429,34 @@ def test_successful_user(self):
...
+    def test_successful_user_not_allowed_to_restricted_paths (self):
+        self.session.auth = ('username-pass-no-restricted-paths', 'fake-password')
+
+        # Mimic a download
+        url = '{artifacts_url}/download/{container}/'.format(
+            artifacts_url=self.artifacts_url,
+            container=self.container
+        )
+        upload = self.session.get(url)
+        assert upload.status_code == 404
```
Comment (bertranddemiddelaer) [inline tests/end2end/test_simple.py:441]: This is used to validate that a user with no access to restricted paths still has access to other paths.

**Comment by bertranddemiddelaer** [comment]
I left a few comments to put some context around the changes I made. Do not hesitate to reach me if you have more questions @tcarmet.

**Review by tcarmet — APPROVED** [review]

**Comment by tcarmet** [comment]
Everything LGTM, I personally have nothing to add, let's just consider sending this PR to @erwan-b for him to have a look later on. But I still think we can merge this through.

**Pull request merged**